### DISCUSSION TITLE
Can I restart a Mirth Windows service by Mirth java script

### DISCUSSION THREAD

**Discussion by lam0620** [nextgenhealthcare/connect#5093] [General]
Stupid question but
Can I restart a Mirth Windows service inside Mirth java script?

My problem is:
My Mirth read XML files from shared network folders. However, I sometimes got network connection error and my channel nerver reconnnects even the network connection fixed. My channel become work after I restart Mirth service manually.

I want to check network connection and restart Mirth service if the connection fail. 

Can I do?

**Comment by jonbartels** [comment]
You probably can by making system calls and running Mirth as a privileged user but this is not a good idea. Having MC restart itself doesn't fix the underlying problem and would add more system administration headaches.

Can you show how your channel is set up? Do you have MC using SMB directly or does it read from a local disk that Windows mounts from the network?

Showing a screenshot of the file reader settings would be helpful. Please blur any addresses, usernames or passwords.

What error do you get? Copy and paste it `using code tags` so it is easy to read.

**Comment by lam0620** [comment]
Thanks @jonbartels for quick reply

My channel's reading files works well but if remote server (that share folder) restart or network connection fail, reading files will be failed.
I think reading network files should work if network connection be fixed. 
I've just restarted MC service manually, reading files become work.

I'm using Mirth Connect Server 3.8.1. My channel's connector type is javascript Reader/javascript Writer

>Can you show how your channel is set up? Do you have MC using SMB directly or does it read from a local disk that Windows mounts from the network?

I have MC using SMB directly and  this is my all javascript Reader code (is implemented Source Tab):
```
const HOS_CONF = $g('HOS_CONF');
const sharedAuth = HOS_CONF.sharedFolderInfo;
const sharedHost = HOS_CONF.sharedFolderInfo.host;
const sharedPath = HOS_CONF.sharedFolderInfo.path;
// List of XML file will read
var listFilesContent = Lists.list();

try {
	//var username = sharedAuth.domain+'@'+sharedAuth.username;
	var username = sharedAuth.username;

	// If domain exist, login by domain
	if (!isEmpty(sharedAuth.domain)) {
		username = sharedAuth.domain+'@'+sharedAuth.username;
	}
	// Domain user format: DOMAIN@user
	var options = new com.mirth.connect.connectors.file.FileSystemConnectionOptions(false, username, sharedAuth.password, null);
	
	var fileConn = new com.mirth.connect.connectors.file.filesystems.SmbFileConnection(sharedHost, options, 5000);
	// array of file: convert fileList to array to sort oldest files first (FIFO - First-In-First-Out order) to make insert/update correct data
	var arrayFileList = [];

	
	if (fileConn != null) {
		var fileList = fileConn.listFiles(sharedPath, '^(ORM|PHY|MDM|ADT|ORU-RPT)_\\w{2}_\\d{14,100}.xml$', true, false);

		// if a file was found 
		if(fileList.size() > 0) { 			
			for (i = 0; i < fileList.size(); i ++) {
				logger.debug(fileList.get(i).getAbsolutePath());
				// Push to array
				arrayFileList.push({ name: fileList.get(i).getName(), lastModified: fileList.get(i).getLastModified()});
			}
		}
		
		arrayFileList.sort(compare);

		if(fileList.size() > 0) {
			for (i = 0; i < arrayFileList.length; i ++) {
				var file = arrayFileList[i];
				logger.debug(file.name);

		    		// Read it's content.
		    		var inputStream = fileConn.readFile(file.name, sharedPath, null);

		    		var contents = org.apache.commons.io.IOUtils.toString(inputStream,'UTF-8');

		    		inputStream.close(); // close this to able to delete in Destination
		
		    		// Remove BOM char (?) at first of content if has
		    		contents = contents.replaceAll("\\A\uFEFF", ""); 

		    		// Add file name to content for deletation use later (at Destination)
		    		var newContents = new XML(contents);
		    		newContents.FILE_NAME = file.name;
		
		
				logger.debug('[Source] contents1: '+ XmlUtil.prettyPrint(newContents));		
	
				listFilesContent.add(newContents);					
			}

			logger.info(listFilesContent);
		}					
	}

} catch (e) {
	logger.error('[Source] Error: '+e.message);
	logger.error(e);

	// Send email error to tracker
	sendMail(e.message);
		
} finally {
	fileConn.destroy();
}

// Return to Destination process
return listFilesContent;

// Compare for sort
function compare(a, b){	
	var comparison = 0;
	if (Number(a.lastModified) > Number(b.lastModified)) {
		comparison = 1;
	} else if (Number(a.lastModified) < Number(b.lastModified)) {
		comparison = -1;
	}
	return comparison;
}
```

I got error message via mail (at sendMail() in catch() block ):

[ERROR] com.mirth.connect.connectors.file.FileConnectorException: Error listing files in dir [his_ris] for pattern [^(ORM|PHY|MDM|ADT|ORU-RPT)_\w{2}_\d{14,100}.xml$]

I appreciate if you help me to find out the cause and how to fix it.

**Comment by pacmano1** [comment]
You may find mounting the file system locally to the mirth server makes more sense - and then testing for the existence of some file that should always exists out there.   Windows can have the file system mounted even when logged off, it is just a bit more painful than in Linux.   Also the service user for mirth should be a named user not the default the installer uses.

**Comment by jonbartels** [comment]
MC uses the [JCIFS library](https://www.jcifs.org/) for SMB file shares. You can see [MCs source](https://github.com/nextgenhealthcare/connect/blob/e66d1122dbeebfcf5a8939beb93d720b2d5071fb/server/src/com/mirth/connect/connectors/file/filesystems/SmbFileConnection.java), make sure you select the right tag version for your MC version.

If restarting MC allows the service to reconnect that seems like there is something being cached, stuck, or re-used inside MC or JCIFS. Maybe like this [JCIFS mailing list message](https://jcifs.samba.narkive.com/jNCXMYFH/reconnect-issue)

I also think that your code should have try/finally blocks to close things like input streams and file handles. 

I agree with pacmano, letting the OS handle netmounts so MC can treat it as a local file is usually better. It is more robust. There are also more sysadmins who can help with a Windows netmount than with MC, you can let a teammate fix the problem! ha ha!

**Comment by pacmano1** [comment]
Curious - why are you using as JS reader here and not the standard file reader?   I am sure it is related to your use case, but I am interested to know why.

**Comment by tonygermano** [comment]
Mounts in windows are all user-based rather than system-based. When I do need to mount a network share in windows I do it in a mirth deploy script, (by running `net use ...`)  which will always mount as the user the mirth server is running as.

**Comment by lam0620** [comment]
Thanks @pacmano1 and @jonbartels 

> Also the service user for mirth should be a named user not the default the installer uses.

You mean MC service user = user who can access the network folder, right?

>Curious - why are you using as JS reader here and not the standard file reader? I am sure it is related to your use case, but I am interested to know why.

Yes! I need to read following FIFO - First-In-First-Out) to make sure correct data processed.

**Comment by lam0620** [comment]
> Mounts in windows are all user-based rather than system-based. When I do need to mount a network share in windows I do it in a mirth deploy script, (by running `net use ...`) which will always mount as the user the mirth server is running as.

Thanks @tonygermano 

I'll try your solution. We must mount (using  `net use ...`) with MC user (`/user: ... `), right?
